flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/SpecificInputTypeStrategies.java

```diff
@@ -46,6 +56,26 @@
     public static final InputTypeStrategy CURRENT_WATERMARK =
             new CurrentWatermarkInputTypeStrategy();
 
+    /**
+     * Input strategy for {@link BuiltInFunctionDefinitions#JSON_OBJECT}.
+     *
+     * <p>The first argument defines the on-null behavior and is followed by any number of key-value
+     * pairs. Keys must be character string literals, while values are arbitrary expressions.
+     */
+    public static final InputTypeStrategy JSON_OBJECT =
+            compositeSequence()
+                    .argument(symbol(JsonOnNull.class))
+                    .finishWithVarying(
+                            repeatingSequence(
+                                    and(logical(LogicalTypeFamily.CHARACTER_STRING), LITERAL),
+                                    or(
+                                            logical(LogicalTypeFamily.CHARACTER_STRING),
+                                            logical(LogicalTypeFamily.BINARY_STRING),
+                                            logical(LogicalTypeFamily.TIMESTAMP),
+                                            logical(LogicalTypeFamily.CONSTRUCTED),
```

Comment:
one last comment: shall we allow `STRUCTURED_TYPE` as well? Internally, this is represented as `RowData` so ideally we only need to add an additional case there and can convert Java POJOs to JSON, which is pretty neat.

Reply:
Would love to. Unfortunately it seems `fromValues` doesn't play nice with this either, it calls `objectToExpression` without the declared type (despite the type being available) which then runs into an error. Working around this with a UDF doesn't work either this time.

I'll make a note for this in the JSON story so we can proceed for now.

Reply:
Could you eloborate this in the JIRA issue? A UDF should work.